### PR TITLE
Added a new FloatRotation3D together with a few needed functions.

### PR DESCRIPTION
**Pull request by niklasborwell**

### CODE REVIEW

File: source/math/FloatPoint3D.ooc [hunk 1]
```diff
@@ -80,3 +81,4 @@ operator * (left: Float, right: FloatPoint3D) -> FloatPoint3D { FloatPoint3D new
 operator / (left: Float, right: FloatPoint3D) -> FloatPoint3D { FloatPoint3D new(left / right x, left / right y, left / right z) }
 operator * (left: Int, right: FloatPoint3D) -> FloatPoint3D { FloatPoint3D new(left * right x, left * right y, left * right z) }
 operator / (left: Int, right: FloatPoint3D) -> FloatPoint3D { FloatPoint3D new(left / right x, left / right y, left / right z) }
+operator - (left: Float, right: FloatPoint3D) -> FloatPoint3D {	FloatPoint3D new(left - right x, left - right y, left - right z) }
```
Comment (davidhesselbom) [inline source/math/FloatPoint3D.ooc]: There's a tab midline here.

File: source/math/FloatRotation3D.ooc [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+			return rz * rx * ry // Yaw -> Pitch -> Roll
+	}
+	round: func -> This { This new(this x round(), this y round(), this z round()) }
+	ceiling: func -> This { This new(this x ceil(), this y ceil(), this z ceil()) }
+	floor: func -> This { This new(this x floor(), this y floor(), this z floor()) }
+	minimum: func (ceiling: This) -> This { This new(this x minimum(ceiling x), this y minimum(ceiling y), this z minimum(ceiling z)) }
+	maximum: func (floor: This) -> This { This new(this x maximum(floor x), this y maximum(floor y), this z maximum(floor z)) }
+	clamp: func ~point(floor, ceiling: This) -> This { This new(this x clamp(floor x, ceiling x), this y clamp(floor y, ceiling y), this z clamp(floor z, ceiling z)) }
+	clamp: func ~float(floor, ceiling: Float) -> This { This new(this x clamp(floor, ceiling), this y clamp(floor, ceiling), this z clamp(floor, ceiling)) }
+	operator + (other: This) -> This { This new(this x + other x, this y + other y, this z + other z) }
```
Comment (davidhesselbom) [inline source/math/FloatRotation3D.ooc]: This doesn't look like it's going to give you the right result...

File: source/math/FloatRotation3D.ooc [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import FloatExtension
+import FloatPoint3D
+import FloatTransform2D
+
+FloatRotation3D: cover {
+	x, y, z: Float
+	init: func@ ~full(=x, =y, =z)
+	init: func@ ~default { this init(0.0f, 0.0f, 0.0f) }
+	init: func@ ~fromPoint(point: FloatPoint3D) { this init(point x, point y, point z) }
+	getTransform: func(k: Float) -> FloatTransform2D {
```
Comment (davidhesselbom) [inline source/math/FloatRotation3D.ooc]: What is `k`?

Comment (davidhesselbom) [inline source/math/FloatRotation3D.ooc]: I see now it already exists in `FloatTransform2D`. But what is it?

File: source/math/FloatRotation3D.ooc [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+			rx := FloatTransform2D createXRotation(this x, k)
+			ry := FloatTransform2D createYRotation(this y, k)
+			rz := FloatTransform2D createZRotation(this z)
+			return rz * rx * ry // Yaw -> Pitch -> Roll
```
Comment (davidhesselbom) [inline source/math/FloatRotation3D.ooc]: Indentation...

Also, have you verified this is correct? A test would be nice.

File: source/math/FloatRotation3D.ooc [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	maximum: func (floor: This) -> This { This new(this x maximum(floor x), this y maximum(floor y), this z maximum(floor z)) }
+	clamp: func ~point(floor, ceiling: This) -> This { This new(this x clamp(floor x, ceiling x), this y clamp(floor y, ceiling y), this z clamp(floor z, ceiling z)) }
+	clamp: func ~float(floor, ceiling: Float) -> This { This new(this x clamp(floor, ceiling), this y clamp(floor, ceiling), this z clamp(floor, ceiling)) }
+	operator + (other: This) -> This { This new(this x + other x, this y + other y, this z + other z) }
+	operator - (other: This) -> This { This new(this x - other x, this y - other y, this z - other z) }
+	operator - -> This { This new(-this x, -this y, -this z) }
+	operator * (other: This) -> This { This new(this x * other x, this y * other y, this z * other z) }
+	operator / (other: This) -> This { This new(this x / other x, this y / other y, this z / other z) }
+	operator * (other: Float) -> This { This new(this x * other, this y * other, this z * other) }
+	operator / (other: Float) -> This { This new(this x / other, this y / other, this z / other) }
```
Comment (davidhesselbom) [inline source/math/FloatRotation3D.ooc]: Unless you know what you're doing with this and didn't just copy it from `FloatPoint3D` or something, I would recommend removing all of these arithmetic functions. What does it even mean to multiply one `FloatRotation3D` with another?

File: source/math/FloatRotation3D.ooc [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import math
+import FloatExtension
+import FloatPoint3D
+import FloatTransform2D
+
+FloatRotation3D: cover {
+	x, y, z: Float
+	init: func@ ~full(=x, =y, =z)
+	init: func@ ~default { this init(0.0f, 0.0f, 0.0f) }
+	init: func@ ~fromPoint(point: FloatPoint3D) { this init(point x, point y, point z) }
```
Comment (davidhesselbom) [inline source/math/FloatRotation3D.ooc]: At some point, we should agree on one way to do this type of thing and not have some classes do `init: func@ ~fromOtherType` and some classes do `toOtherType: func`. Right now, we're doing both.

File: source/math/FloatRotation3D.ooc (hunk 5 above)
Comment (niklasborwell) [inline source/math/FloatRotation3D.ooc]: I need this in some of the functions that I use.

File: source/math/FloatRotation3D.ooc (hunk 4 above)
Comment (niklasborwell) [inline source/math/FloatRotation3D.ooc]: This has been used for a long time. I just moved it here from quaternion.

File: source/math/FloatRotation3D.ooc (hunk 6 above)
Comment (niklasborwell) [inline source/math/FloatRotation3D.ooc]: Personally I would say fromOtherType feels more natural and toOtherType should be removed. I've been avoiding toOtherType recently.

**Comment by davidhesselbom** [comment]
Today is your lucky day! Rebase and I'll merge it.

File: source/math/FloatRotation3D.ooc [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+		rx := FloatTransform2D createXRotation(this x, zDistance)
+		ry := FloatTransform2D createYRotation(this y, zDistance)
+		rz := FloatTransform2D createZRotation(this z)
+		return rz * rx * ry // Yaw -> Pitch -> Roll
+	}
+	round: func -> This { This new(this x round(), this y round(), this z round()) }
+	ceiling: func -> This { This new(this x ceil(), this y ceil(), this z ceil()) }
+	floor: func -> This { This new(this x floor(), this y floor(), this z floor()) }
+	minimum: func (ceiling: This) -> This { This new(this x minimum(ceiling x), this y minimum(ceiling y), this z minimum(ceiling z)) }
+	maximum: func (floor: This) -> This { This new(this x maximum(floor x), this y maximum(floor y), this z maximum(floor z)) }
```
Comment (davidhesselbom) [inline source/math/FloatRotation3D.ooc]: This still needs fixing, unless you've done it already.

File: source/math/FloatRotation3D.ooc [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	operator < (other: This) -> Bool { this x < other x && this y < other y && this z < other z }
+	operator > (other: This) -> Bool { this x > other x && this y > other y && this z > other z}
+	operator <= (other: This) -> Bool { this x <= other x && this y <= other y && this z <= other z }
+	operator >= (other: This) -> Bool { this x >= other x && this y >= other y && this z >= other z }
+	toString: func -> String { "%.8f" formatFloat(this x) >> ", " & "%.8f" formatFloat(this y) >> ", " & "%.8f" formatFloat(this z) }
+}
+operator * (left: Float, right: FloatRotation3D) -> FloatRotation3D { FloatRotation3D new(left * right x, left * right y, left * right z) }
+operator / (left: Float, right: FloatRotation3D) -> FloatRotation3D { FloatRotation3D new(left / right x, left / right y, left / right z) }
+operator * (left: Int, right: FloatRotation3D) -> FloatRotation3D { FloatRotation3D new(left * right x, left * right y, left * right z) }
+operator / (left: Int, right: FloatRotation3D) -> FloatRotation3D { FloatRotation3D new(left / right x, left / right y, left / right z) }
```
Comment (davidhesselbom) [inline source/math/FloatRotation3D.ooc]: And these were already removed, I think.

Comment (niklasborwell) [inline source/math/FloatRotation3D.ooc]: Thanks for the information. As you probably have realized, it hasn't been pushed yet.

**Comment by niklasborwell** [comment]
Ok, both requests now updated, rebased, tested and memtested.

**Pull request merged**